### PR TITLE
multi delegate contract

### LINKED ISSUE #13: Write a contract that allows delegating ENS tokens selectively to delegates.

### PR DESCRIPTION
**Pull request by mdtanrikulu**
resolves #13

### CODE REVIEW

File: contracts/ENSGovernor.sol [hunk 1]
```diff
@@ -46,7 +47,7 @@ contract ENSGovernor is Governor, GovernorProposalThreshold, GovernorCountingSim
     function getVotes(address account, uint256 blockNumber)
         public
         view
-        override(IGovernor, GovernorVotes)
+        override(IGovernor, Governor)
```
Comment (Arachnid) [inline contracts/ENSGovernor.sol:50]: What's the motivation behind the changes to this file?

File: contracts/ERC20MultiDelegate.sol [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/token/ERC1155/ERC1155.sol";
+import "@openzeppelin/contracts/token/ERC20/extensions/ERC20Votes.sol";
+
+/**
+ * @dev A child contract which will be deployed by the ERC20MultiDelegate utility contract
+ * This is a proxy delegator contract to vote given delegatee on behalf of original delegator
+ */
+contract ERC20ProxyDelegator {
+    constructor(ERC20Votes _token, address _delegatee) {
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:15]: Weren't we going to have this selfdestruct as well, to save gas?

File: contracts/ERC20MultiDelegate.sol [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+contract ERC20MultiDelegate is ERC1155, Ownable {
+    using Address for address;
+
+    ERC20Votes token;
+
+    /**
+     * @dev Constructor.
+     * @param _token The ERC20 token address
+     */
+    constructor(ERC20Votes _token) ERC1155("http://some.metadata.url/{id}") {
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:33]: Maybe take the metadata URL as a constructor argument too?

File: contracts/ERC20MultiDelegate.sol [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        token = _token;
+    }
+
+    /**
+     * @dev Public method for the delegation of multiple delegatees.
+     * @param delegatees List of delegatee addresses
+     * @param amounts ERC20 voting power amount to be distributed among delegatees
+     */
+    function delegateMulti(
+        address[] calldata delegatees,
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:43]: An array of structs would probably be cleaner here.

File: contracts/ERC20MultiDelegate.sol [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+            delegatees.length == amounts.length,
+            "Amounts should be defined for each delegatee"
+        );
+
+        uint256[] memory ids = new uint256[](delegatees.length);
+
+        for (uint256 index = 0; index < delegatees.length; ) {
+            address delegatee = delegatees[index];
+            // creates a proxy delegator contract with deterministic address
+            // salt occurs from delegatee address + sender address
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:60]: I thought the idea was to have a single proxy contract for each delegatee, not for each (sender, delegatee)? That way we only need to create them once, not once per user.

File: contracts/ERC20MultiDelegate.sol [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+            // transfer ERC20 for the proxy delegation
+            // initialize the contract after with predetermined address
+            token.transferFrom(
+                msg.sender,
+                predeterminedProxyAddress,
+                amounts[index]
+            );
+            new ERC20ProxyDelegator{salt: salt}(
+                token,
+                delegatee
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:76]: Constructor arguments don't need to be part of the salt value, because they comprise part of the constructor bytecode.

File: contracts/ERC20MultiDelegate.sol [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+            "Amounts should be defined for each delegatee"
+        );
+
+        uint256[] memory ids = new uint256[](delegatees.length);
+
+        for (uint256 index = 0; index < delegatees.length; ) {
+            address delegatee = delegatees[index];
+            // creates a proxy delegator contract with deterministic address
+            // salt occurs from delegatee address + sender address
+            bytes memory bytecode = getBytecode(token, delegatee);
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:61]: `getBytecode` and `getAddress` should probably be combined into a single utility function.

File: contracts/ERC20MultiDelegate.sol [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+    constructor(ERC20Votes _token) ERC1155("http://some.metadata.url/{id}") {
+        token = _token;
+    }
+
+    /**
+     * @dev Public method for the delegation of multiple delegatees.
+     * @param delegatees List of delegatee addresses
+     * @param amounts ERC20 voting power amount to be distributed among delegatees
+     */
+    function delegateMulti(
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:42]: Either this should be called `deposit` or `withdraw` should be called `undelegateMulti`. Personally I think `depositMulti/withdrawMulti` makes sense, with corresponding convenience functions that only deposit and assign to a single address.

File: contracts/ERC20MultiDelegate.sol [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+            // get the delegatee list from user
+            address delegatee = delegatees[index];
+            // PDT - proxy delegation token
+            // check if user has the PDT for the provided delegatees
+            uint256 amount = ERC1155(this).balanceOf(
+                msg.sender,
+                uint256(uint160(delegatee))
+            );
+            // burn PDT's
+            _burn(msg.sender, uint256(uint160(delegatee)), amount);
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:101]: Should probably use `burnBatch` here, and put it at the end to avoid reentrancy concerns.

File: test/delegatemulti.js [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+    multiDelegate = await ENSMultiDelegate.deploy(token.address);
+    await multiDelegate.deployed();
+
+    await registry.setSubnodeOwner(ROOT_NODE, labelHash, deployer);
+    await registry.setResolver(node, resolver.address);
+
+    await increaseTime(365 * 24 * 60 * 60);
+    const mintAmount = (await token.totalSupply()).div(50);
+    await token.mint(deployer, mintAmount);
+    const avatar = 'some avatar';
```
Comment (Arachnid) [inline test/delegatemulti.js:60]: I don't think this is necessary?

File: contracts/ERC20MultiDelegate.sol [hunk 11]
```diff
@@ -0,0 +1,150 @@
+// SPDX-License-Identifier: MIT
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:1]: An 'exchange' method that swaps tokens from one delegate to another would be nice to have, too, instead of having to withdraw and re-deposit.

File: test/delegatemulti.js [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+    await multiDelegate.delegateMulti(
+      delegatees,
+      delegatees.map((_) => delegatorTokenAmount.div(delegatees.length))
+    );
+
+    const delegatorTokenAmountAfter = await token.balanceOf(deployer);
+    expect(delegatorTokenAmountAfter.toString()).to.equal('0');
+
+
+    // delegator must have 1/2 of the votes the delegator delegated
```
Comment (Arachnid) [inline test/delegatemulti.js:92]:
```suggestion
    // delegator must have 1/4 of the votes the delegator delegated
```

File: test/delegatemulti.js [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+    const votesOfDelegator = await token.getVotes(deployer);
+    expect(votesOfDelegator.toString()).to.equal(
+      delegatorTokenAmount.div(delegatees.length).toString()
+    );
+
+    // delegatee must have 1/2 of the votes the delegator delegated
```
Comment (Arachnid) [inline test/delegatemulti.js:98]:
```suggestion
    // delegatee must have 1/4 of the votes the delegator delegated
```

File: test/delegatemulti.js [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+    // give allowance to multi delegate contract
+    await token.approve(multiDelegate.address, delegatorTokenAmount);
+
+    const delegatees = [];
+    await expect(
+      multiDelegate.delegateMulti(delegatees, [delegatorTokenAmount])
+    ).to.be.revertedWith('You should pick at least one delegatee');
+  });
+
+  it('contract should revert if no amount provided', async () => {
```
Comment (Arachnid) [inline test/delegatemulti.js:153]: There should be tests for insufficient balances too - either depositing or withdrawing.

**Review by Arachnid — CHANGES_REQUESTED** [review]

File: contracts/ERC20MultiDelegate.sol [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+    }
+}
+
+/**
+ * @dev A utility contract to let delegators to pick multiple delegatee
+ */
+contract ERC20MultiDelegate is ERC1155, Ownable {
+    using Address for address;
+
+    ERC20Votes token;
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:27]: This should be public.

File: contracts/ENSGovernor.sol (hunk 1 above)
Comment (mdtanrikulu) [reply contracts/ENSGovernor.sol:50]: Compiler complains;
```
Function needs to specify overridden contract "Governor".
Invalid contract specified in override list: "GovernorVotes".
```

File: contracts/ERC20MultiDelegate.sol (hunk 2 above)
Comment (mdtanrikulu) [reply contracts/ERC20MultiDelegate.sol:15]: after our last discussion as I remember the conclusion was; we do not need it.

File: contracts/ERC20MultiDelegate.sol (hunk 5 above)
Comment (mdtanrikulu) [reply contracts/ERC20MultiDelegate.sol]: updated

File: contracts/ERC20MultiDelegate.sol (hunk 2 above)
Comment (Arachnid) [reply contracts/ERC20MultiDelegate.sol:15]: I wonder if it would be worth the extra gas to store the token, delegatee, and msg.sender values as immutable public variables?

File: contracts/ERC20MultiDelegate.sol [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+
+        burnBatch(msg.sender, sourceIds, amounts);
+        mintBatch(msg.sender, targetIds, amounts);
+    }
+
+    /**
+     * @dev Withdraws delegated ERC20 voting power from proxy delegators to the actual delegator
+     * @param delegatees List of delegatee addresses
+     */
+    function withdrawMulti(address[] calldata delegatees) external {
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:120]: This should probably allow partial withdraws by taking an array of amounts as well.

File: contracts/ERC20MultiDelegate.sol [hunk 17]
```diff
@@ -0,0 +1,251 @@
...
+    }
+
+    function deployProxyDelegatorIfNeeded(
+        address delegatee
+    ) internal returns (address) {
+        (address proxyAddress, bytes32 salt) = retrieveProxyContractAddress(
+            token,
+            delegatee
+        );
+        new ERC20ProxyDelegator{salt: salt}(token, delegatee);
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:173]: This appears to deploy unconditionally, not only if needed.

File: contracts/ERC20MultiDelegate.sol [hunk 18]
```diff
@@ -0,0 +1,251 @@
...
+        return proxyAddress;
+    }
+
+    function getBalanceForDelegatee(
+        address delegatee
+    ) internal view returns (uint256) {
+        return ERC1155(this).balanceOf(msg.sender, uint256(uint160(delegatee)));
+    }
+
+    function getProxyContractAddress(
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:183]: Why do we have this function wrapping the othe rone?

File: contracts/ERC20MultiDelegate.sol [hunk 19]
```diff
@@ -0,0 +1,251 @@
...
+            "Failed to transfer voting power"
+        );
+    }
+
+    function mintBatch(
+        address account,
+        uint256[] memory ids,
+        uint256[] memory amounts
+    ) internal {
+        _mintBatch(account, ids, amounts, "");
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:209]: Are these functions necessary?

File: tasks/maketree.js [hunk 20]
```diff
@@ -0,0 +1,57 @@
+const { task } = require('hardhat/config');
```
Comment (Arachnid) [inline tasks/maketree.js:1]: Why is this being added?

File: test/delegatemulti.js [hunk 21]
```diff
@@ -0,0 +1,227 @@
...
+    const delegatees = [];
+    const delegateeAmountArray = delegatees.map((delegatee) => ([
+      delegatee,
+      delegatorTokenAmount.div(delegatees.length)
+    ]));
+    await expect(
+      multiDelegate.depositMulti(delegateeAmountArray)
+    ).to.be.revertedWith('You should pick at least one delegatee');
+  });
+});
```
Comment (Arachnid) [inline test/delegatemulti.js:227]: This needs to add tests for delegting to and transferring to an account that already has a proxy contract.

File: contracts/ERC20MultiDelegate.sol [hunk 22]
```diff
@@ -0,0 +1,251 @@
...
+
+        mintBatch(msg.sender, ids, amounts);
+    }
+
+    /**
+    * @dev Re-delegates voting power between delegatees.
+    * @param delegateePairs The list of current delegatee addresses (from which the voting power is withdrawn) 
+    * and target delegatee addresses (which the voting power will be delegated).
+    */
+    function reDeposit(SourceTargetDelegatee[] calldata delegateePairs) external {
```
[the remark this answers is not in the document]
Comment (mdtanrikulu) [reply contracts/ERC20MultiDelegate.sol:86]: One thing I am unsure about this method, should we map the transfer as 1:1, so for each `source` we will require a `target` address. or should it give freedom to user so the delegation can be many to many, like; for 2 `source` address, user can define let's say 3 `target` address with arbitrary `amount` to split or move the voting power?

File: contracts/ERC20MultiDelegate.sol (hunk 19 above)
Comment (mdtanrikulu) [reply contracts/ERC20MultiDelegate.sol:209]: moved it into a separate method for modularity, in case we may need to perform additional checks or calls. can just call the private method instead if it seem unnecessary.

File: contracts/ERC20MultiDelegate.sol (hunk 18 above)
Comment (mdtanrikulu) [reply contracts/ERC20MultiDelegate.sol:183]: oh, leftover. going to update it.

File: contracts/ERC20MultiDelegate.sol (hunk 2 above)
Comment (mdtanrikulu) [reply contracts/ERC20MultiDelegate.sol:15]: | depositMulti | none | public | public immutable |
| ---- | --- | --- | --- |
| min | 448520 | 766173 | 631925 | 
| max | 864048 | 1499405 | 1230909 |

File: tasks/maketree.js (hunk 20 above)
Comment (mdtanrikulu) [reply tasks/maketree.js:1]: the import of the function was missing on line 4

File: contracts/ERC20MultiDelegate.sol (hunk 17 above)
Comment (mdtanrikulu) [reply contracts/ERC20MultiDelegate.sol]: fixed

File: contracts/ERC20MultiDelegate.sol (hunk 22 above)
[the remark this answers is not in the document]
Comment (mdtanrikulu) [reply contracts/ERC20MultiDelegate.sol]: okay I change it as I described above, but instead of three different arrays, I required user to provide two array of structs;
```
[[source_address, withdraw_amount] ...], [[target_address, target_amount]...]
```
both array sizes can be arbitrary so user can redeposit from many to many. 

In case user choose to withdraw more amount from source addresses in total but deposit less to targets in total, the remaining amount will be reimbursed to the caller.

File: contracts/ERC20MultiDelegate.sol (hunk 2 above)
Comment (Arachnid) [reply contracts/ERC20MultiDelegate.sol:15]: An extra 50%! Let's leave it out, then.

File: contracts/ERC20MultiDelegate.sol [hunk 23]
```diff
@@ -0,0 +1,251 @@
...
+/**
+ * @dev A utility contract to let delegators to pick multiple delegatee
+ */
+contract ERC20MultiDelegate is ERC1155, Ownable {
+    using Address for address;
+
+    ERC20Votes public token;
+
+    struct DelegateeAmount {
+        address delegatee;
```
[the remark this answers is not in the document]
Comment (Arachnid) [reply contracts/ERC20MultiDelegate.sol:30]: Ping!

File: contracts/ERC20MultiDelegate.sol (hunk 2 above)
Comment (Arachnid) [reply contracts/ERC20MultiDelegate.sol:15]: Worth investigating whether writing this in Yul would save a lot of gas, too.

**Review by Arachnid — CHANGES_REQUESTED** [review]

File: contracts/ERC20MultiDelegate.sol [hunk 24]
```diff
@@ -0,0 +1,434 @@
...
+    constructor(ERC20Votes _token, address _delegatee) {
+        _token.approve(msg.sender, type(uint256).max);
+        _token.delegate(_delegatee);
+    }
+}
+
+/**
+ * @dev A utility contract to let delegators to pick multiple delegatee
+ */
+contract ERC20MultiDelegate is ERC1155, Ownable {
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:24]: This should probably be renamed; the name implies it's an ERC20. Maybe `MultiDelegateToken`?

File: contracts/ERC20MultiDelegate.sol [hunk 25]
```diff
@@ -0,0 +1,434 @@
...
+        mintBatch(msg.sender, ids, amounts);
+    }
+
+    /**
+     * @dev Re-delegates voting power between delegatees.
+     * @param sources The list of source delegatee addresses and amounts to withdraw.
+     * @param targets The list of target delegatee addresses and amounts to redeposit.
+     * The remaning part if any, will be withdrawn to the user account
+     */
+    function reDeposit(
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:94]: I feel like all three methods could be replaced by one, with a list of source and destination tokens, and just have it transfer from or to the ERC20 if there's not enough on one side or the other.

File: contracts/ERC20MultiDelegate.sol [hunk 26]
```diff
@@ -0,0 +1,434 @@
...
+    /**
+     * @dev Executes the re-deposit process for multiple source and target delegatees.
+     * @param sources The list of source delegatees with the amounts to withdraw.
+     * @param targets The list of target delegatees with the amounts to re-deposit.
+     * @param record The current state of the re-deposit process.
+     */
+    function _reDeposit(
+        DelegateeAmount[] calldata sources,
+        DelegateeAmount[] calldata targets,
+        ReDepositRecord memory record
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:143]: I think you should integrate the outer loop into this function, so the internal state doesn't need to be preserved in a struct.

**Review by Arachnid — CHANGES_REQUESTED** [review]

File: contracts/ERC20MultiDelegate.sol (hunk 6 above)
Comment (mdtanrikulu) [reply contracts/ERC20MultiDelegate.sol]: updated;
https://github.com/ensdomains/governance/blob/4e4c7cbc89d82aa9f035a7abcc7a1854432eb423/contracts/ERC20MultiDelegate.sol#L431-L432

File: contracts/ERC20MultiDelegate.sol (hunk 16 above)
Comment (mdtanrikulu) [reply contracts/ERC20MultiDelegate.sol]: updated: 

https://github.com/ensdomains/governance/blob/4e4c7cbc89d82aa9f035a7abcc7a1854432eb423/contracts/ERC20MultiDelegate.sol#L29-L32

https://github.com/ensdomains/governance/blob/4e4c7cbc89d82aa9f035a7abcc7a1854432eb423/contracts/ERC20MultiDelegate.sol#L290

File: contracts/ERC20MultiDelegate.sol (hunk 15 above)
Comment (mdtanrikulu) [reply contracts/ERC20MultiDelegate.sol]: updated https://github.com/ensdomains/governance/blob/4e4c7cbc89d82aa9f035a7abcc7a1854432eb423/contracts/ERC20MultiDelegate.sol#L27

File: test/delegatemulti.js (hunk 14 above)
Comment (mdtanrikulu) [reply test/delegatemulti.js]: deposits addressed here;

https://github.com/ensdomains/governance/blob/4e4c7cbc89d82aa9f035a7abcc7a1854432eb423/test/delegatemulti.js#L619

partially here;
https://github.com/ensdomains/governance/blob/4e4c7cbc89d82aa9f035a7abcc7a1854432eb423/test/delegatemulti.js#L419

withdraw here;
https://github.com/ensdomains/governance/blob/4e4c7cbc89d82aa9f035a7abcc7a1854432eb423/test/delegatemulti.js#L533-L534

File: contracts/ERC20MultiDelegate.sol (hunk 24 above)
Comment (mdtanrikulu) [reply contracts/ERC20MultiDelegate.sol:24]: I actually prefixed as ERC20 in purpose. Since it's supports any token with `ERC20Votes`. But if you think it's misleading, I am also fine with `MultiDelegateToken`.

File: contracts/ERC20MultiDelegate.sol (hunk 23 above)
[the remark this answers is not in the document]
Comment (mdtanrikulu) [reply contracts/ERC20MultiDelegate.sol]: updated: [98b4710](https://github.com/ensdomains/governance/pull/16/commits/98b47106b0561256cae9db88be91c51e4d1133bb)

File: contracts/ERC20MultiDelegate.sol (hunk 3 above)
Comment (mdtanrikulu) [reply contracts/ERC20MultiDelegate.sol]: here: https://github.com/ensdomains/governance/blob/4e4c7cbc89d82aa9f035a7abcc7a1854432eb423/contracts/ERC20MultiDelegate.sol#L51-L54

File: contracts/ERC20MultiDelegate.sol (hunk 24 above)
Comment (Arachnid) [reply contracts/ERC20MultiDelegate.sol:24]: My concern is that named as it is it strongly implies it's an ERC20 token.

File: contracts/ERC20MultiDelegate.sol (hunk 19 above)
Comment (Arachnid) [reply contracts/ERC20MultiDelegate.sol:392]: I think unless it needs to be called from multiple places, you should skip having a separate internal method.

File: contracts/ERC20MultiDelegate.sol (hunk 2 above)
Comment (mdtanrikulu) [reply contracts/ERC20MultiDelegate.sol]: keeping this conversation open for future improvements

File: contracts/ERC20MultiDelegate.sol [hunk 27]
```diff
@@ -0,0 +1,251 @@
...
+    /**
+     * @dev Deposits and delegates voting power to the multiple delegatees.
+     * @param delegateeAmounts The list of delegatee addresses and corresponding list of ERC20 voting power amount amounts to delegate.
+     */
+    function depositMulti(
+        DelegateeAmount[] calldata delegateeAmounts
+    ) external {
+        uint256 delegateesLength = delegateeAmounts.length;
+
+        require(
```
[the remark this answers is not in the document]
Comment (mdtanrikulu) [reply contracts/ERC20MultiDelegate.sol:78]: do you think same applies for this too? 

https://github.com/ensdomains/governance/blob/a96d7b921dd3cf86b73afd6e76204933d37adb2c/contracts/ERC20MultiDelegate.sol#L78-L81

**Review by zardokofficial2023 — APPROVED** [review]

**Review by Arachnid — CHANGES_REQUESTED** [review]

File: contracts/ERC20MultiDelegate.sol [hunk 28]
```diff
@@ -0,0 +1,298 @@
...
+     * @param targets The list of target delegates with the amounts to deposit.
+     */
+    function delegateMulti(
+        DelegateAmount[] calldata sources,
+        DelegateAmount[] calldata targets
+    ) external {
+        _delegateMulti(sources, targets);
+    }
+
+    function _delegateMulti(
```
[the remark this answers is not in the document]
Comment (mdtanrikulu) [reply contracts/ERC20MultiDelegate.sol:64]: - I moved the redistribution strategy out of the contract now, and the input is just as you mentioned `address[] sources, address[] targets, uint256[] amounts` but the thing is there is no erc20 batch transfer available + we still need to deploy proxy contracts for each delegate, so I still need to stick doing everything under “while” loop. (except minting/burning 1155 tokens to/from caller, to keep the record of delegates, caller delegated)

- And not sure if I understood this statement “If one list or the other is longer, mint or burn tokens to/from the caller” cuz it’s not only about minting/burning erc1155 tokens but also using message sender as a source or target to transfer voting power, if one list is shorter than other. so this again done under “while” as I mentioned above.

- Also an additional complexity comes from ERC1155 mintBatch/burnBatch. using delegate address as a token id does not work, expected id type is uint256, so during transfers under “while” loop I cast all delegate addresses to uint256 and store in the array, then using it for batch minting/burning processes.

- here is the updated version: https://github.com/ensdomains/governance/blob/22a3ded95789134d7693dbbc72db411a5ae3d1a8/contracts/ERC20MultiDelegate.sol#L56

A small note that; surprisingly, this change didn't really affect gas usage, the minimum gas usage was somewhere around "~92500", now it's "~91500".

**Review by zardokofficial2023 — APPROVED** [review]

File: contracts/ERC20MultiDelegate.sol [hunk 29]
```diff
@@ -0,0 +1,245 @@
...
+    function delegateMulti(
+        address[] calldata sources,
+        address[] calldata targets,
+        uint256[] calldata amounts
+    ) external {
+        _delegateMulti(sources, targets, amounts);
+    }
+
+    function _delegateMulti(
+        address[] calldata sources,
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:65]: If you take `uint256` IDs instead of addresses, you could avoid the need to create the `sourceIds` and `targetIds` arrays, and save a lot of gas on that - just cast them back to addresses in the loop.

File: contracts/ERC20MultiDelegate.sol [hunk 30]
```diff
@@ -0,0 +1,245 @@
...
+        );
+
+        uint256 transferIndex = 0;
+        uint256[] memory sourceIds = new uint256[](sourcesLength);
+        uint256[] memory targetIds = new uint256[](targetsLength);
+        uint256[] memory withdrawnAmounts = new uint256[](sourcesLength);
+        uint256[] memory depositedAmounts = new uint256[](targetsLength);
+
+        // Iterate until all source and target delegates have been processed.
+        while (transferIndex < sourcesLength || transferIndex < targetsLength) {
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:84]: This would be easier to read as a for loop.

File: contracts/ERC20MultiDelegate.sol [hunk 31]
```diff
@@ -0,0 +1,245 @@
...
+
+    function _delegateMulti(
+        address[] calldata sources,
+        address[] calldata targets,
+        uint256[] calldata amounts
+    ) internal {
+        uint256 sourcesLength = sources.length;
+        uint256 targetsLength = targets.length;
+
+        require(
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:72]: You should also check that `amounts` has at least as many elements as the larger of `sources` and `targets`.

File: contracts/ERC20MultiDelegate.sol [hunk 32]
```diff
@@ -0,0 +1,245 @@
...
+        }
+    }
+
+    /**
+     * @dev Processes the delegation transfer between a source delegate and a target delegate.
+     * @param source The source delegate from which tokens are being withdrawn.
+     * @param target The target delegate to which tokens are being transferred.
+     * @param amount The amount of tokens transferred between the source and target delegates.
+     */
+    function _processDelegation(
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:138]: Shouldn't this use `createProxyDelegatorAndTransfer`?

File: contracts/ERC20MultiDelegate.sol [hunk 33]
```diff
@@ -0,0 +1,245 @@
...
+
+        // if the proxy contract has not been deployed, deploy it
+        if (bytecode.length == 0) {
+            new ERC20ProxyDelegator{salt: 0}(token, delegate);
+            emit ProxyDeployed(delegate, proxyAddress);
+        }
+        return proxyAddress;
+    }
+
+    function getBalanceForDelegatee(
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:210]: Just `delegate`, not `delegatee`.

**Review by Arachnid — CHANGES_REQUESTED** [review]
Can you run code coverage and add tests as needed, too?

File: contracts/ERC20MultiDelegate.sol (hunk 33 above)
Comment (mdtanrikulu) [reply contracts/ERC20MultiDelegate.sol]: oh I missed it out, renamed now.

File: contracts/ERC20MultiDelegate.sol (hunk 32 above)
Comment (mdtanrikulu) [reply contracts/ERC20MultiDelegate.sol:124]: `createProxyDelegatorAndTransfer` transfers from `msg.sender` to the `target`. it doesn't do check if source (msg.sender) has a proxy contract. I am using `deployProxyDelegatorIfNeeded` & `transferBetweenDelegators` instead.  One improvement here should be the changing the order of methods. 

instead of;
```
transferBetweenDelegators(source, target, amount);
deployProxyDelegatorIfNeeded(target);
```
it must be;
```
deployProxyDelegatorIfNeeded(target);
transferBetweenDelegators(source, target, amount);
```

**Comment by mdtanrikulu** [comment]
> Can you run code coverage and add tests as needed, too?

sure, added it

File: contracts/ERC20MultiDelegate.sol [hunk 34]
```diff
@@ -0,0 +1,227 @@
...
+
+    function deployProxyDelegatorIfNeeded(
+        address delegate
+    ) internal returns (address) {
+        address proxyAddress = retrieveProxyContractAddress(token, delegate);
+
+        // check if the proxy contract has already been deployed
+        bytes memory bytecode;
+        assembly {
+            bytecode := extcodesize(proxyAddress)
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:181]: This returns the *size* of the bytecode, not the actual bytecode. I'm surprised this compiles!

Can you also check why there aren't any failing tests as a result of this?

File: contracts/ERC20MultiDelegate.sol [hunk 35]
```diff
@@ -0,0 +1,227 @@
...
+        return proxyAddress;
+    }
+
+    function getBalanceForDelegate(
+        address delegate
+    ) internal view returns (uint256) {
+        return ERC1155(this).balanceOf(msg.sender, uint256(uint160(delegate)));
+    }
+
+    function getContractAddress(
```
Comment (Arachnid) [inline contracts/ERC20MultiDelegate.sol:198]: Combining these three functions may save gas by reducing copies, and potentially allowing the optimizer to avoid realizing some intermediate results in memory.

File: contracts/ERC20MultiDelegate.sol (hunk 34 above)
Comment (Arachnid) [reply contracts/ERC20MultiDelegate.sol:181]: Right, but the variable is of type `bytes memory` and you're asking for its `length`. It should be a`uint`.

**Review by Arachnid — APPROVED** [review]

**Comment by mdtanrikulu** [comment]
re-targeting branch to staging
